### ISSUE TITLE
Fan and LED Control

### ISSUE THREAD

**Issue by hiroTM** [rogers0/OpenLinkstation#5]
Thank you for a very useful script.
And to challenge several times, I was somehow able to put a Debian.
However, my LS-WVL After starting, LED continue to blink and FAN has rotated with full.

And to control these, Do you have a Web site that will be the good way or reference?

**Comment by rogers0** [comment]
Thanks for your feedback.
Yes, this is on my todo-list, but unfortunately not in high priority.
Maybe you can figure out by your own.

You can refer the [Issue#1](https://github.com/rogers0/OpenLinkstation/issues/1) as reference.
It's for LS-WXL, so you don't need it. But you can understand how to control the GPIO.

**Comment by ozdroid** [comment]
@hiroTM , contact me or get me your EMAIL.... i have the code control fan & led for  LS_WVL .

**Comment by hiroTM** [comment]
On 2015/10/31 11:35, ozdroid wrote:

@hiroTM , contact me or get me your EMAIL.... i have the code control fan & led for  LS_WVL . ---
Reply to this email directly or view it on GitHub:
https://github.com/rogers0/OpenLinkstation/issues/5#issuecomment-152687746 
hi ozdroid,

I'm sorry to reply late.
We want to correct the LED blink, but Do you have any good code?

**Comment by dslam127** [comment]
heroTM, I also have a LS_WVL Pro Duo. The gpio pinouts on this patch match perfectly: http://lists.infradead.org/pipermail/linux-arm-kernel/2015-June/352973.html

I use them with rogers0 comment:
echo {pin} >/sys/class/gpio/export
echo out > /sys/class/gpio/gpio{pin}/direction

then you can change them with
echo {1|0} > /sys/class/gpio/gpio{pin}/value

I haven't entirely figured out how to make the fan speeds variable but I am able to put it in low or high mode. And the power light can be set to steady on instead of blinking. Also, I am able to capture the events from the function and power buttons by using "in" instead of "out" for the direction.

I'm currently working on some scripts for user interaction. Hope this helps. Thank you rogers0 for the debian install. It worked perfectly. The directions were a little hard to follow but they worked.

**Comment by rogers0** [comment]
dslam127, very glad to hear that you worked out the LED/Fan control.
If possible, please put it in github or other public repo. So as the community can all benefit and also contribute.
No problem if there're still some parts not working.

**Comment by dslam127** [comment]
I forked a branch with some instructions and modifications. Note I have only tested this on my LS-WVL/E. Access the fork here https://github.com/dslam127/OpenLinkstation or open up the branches and find it. Good Luck!

**Comment by rogers0** [comment]
dslam127, thanks for your sharing! Let me try it.
Maybe you can send me a pull-request. Or you prefer to be still working on it till a more perfect status? :-P

**Comment by dslam127** [comment]
sorry sir, didn't know i needed to do that. I don't normally use git but
I'm trying to learn. pull-request submitted

**Comment by rogers0** [comment]
dslam127, thanks for your work!

I have updated the DTB/DTS: https://github.com/rogers0/OpenLinkstation/tree/master/1_debootstrap/dtb

Could you help to test whether your code still works or not?
I guess there may need some minor fix, such as 1 <=> 0 change for some LED pins.

Looking forward to your feedback!

**Comment by qakcn** [comment]
I'm using an LS-WXL, and encounter same issue.
I tried `echo XX > /sys/class/gpio/export` and nothing happened.
I tried update DTB and still nothing happened.

The fan is rolling at full speed. Please help.

**Comment by qakcn** [comment]
Finally, I found GPIO pins in your patchset. Thank you.

But still not all is functional. Only fan control can work.
LEDs and the power switch is useless as before.

`ls` result of `/sys/class/gpio/` shows that only `gpio40`(fan status), `gpio47`(fan high) and `gpio48`(fan low) are valid.

Really need help!

**Comment by dslam127** [comment]
scripts are no longer valid after fresh debian install... need to figure
out a new way to map and control the gpio functions.

**Comment by dslam127** [comment]
I don't think my scripts are working after the new update... I'll do some digging and check back with y'all in a couple days after I'm done with final exams. Check back later please...

**Comment by rogers0** [comment]
if you use testing/sid kernel, the DTB included is already latest (same as what I provided here)
and there's no need to "export" anything

for fan control, please check this folder:
- /sys/devices/platform/gpio_fan/hwmon/hwmon0

for temperature, please check this folder:
- /sys/devices/virtual/thermal/thermal_zone0

if you use kernel 4.1 or 4.2, maybe the folder is located elsewhere, but the structure is almost the same

**Comment by qakcn** [comment]
I found these in `/sys/devices/gpio_leds/leds/`:

    linkstation:amber:info
    linkstation:red:alarm
    linkstation:red:hdderr1
    linkstation:blue:function
    linkstation:red:function
    linkstation:blue:power
    linkstation:red:hdderr0

And they works good.

But no fan or thermal can be found in the path you mentioned.

I also found `/sys/devices/gpio_fan` and `/sys/devices/gpio_keys`, but I don't know how it works.

I will continue trying,

**Comment by rogers0** [comment]
@qakcn 
For LS-WXL (kirkwood 6281 series chipset), there's no thermal support
it only exists on kirkwood 6282 series, e.g. LS-VL/WVL/QVL.

For fan, it should be there. Maybe in different path. I confirm folder below works on debian kernel 4.5 / 4.6:
- /sys/devices/platform/gpio_fan/hwmon/hwmon0
if there's no such folder, you can try:
- /sys/devices/gpio_fan/hwmon/hwmon0
or replace hwmon0 with hwmon1, etc..

**Comment by qakcn** [comment]
No any `hwmon` things in the folder. Nor anything useful. I tried everthing there, but none of them is.
